compilers/concrete-compiler/compiler/lib/Dialect/TFHE/Analysis/ExtractStatistics.cpp

```diff
@@ -0,0 +1,141 @@
+  mlir::Value startValue = op.getLowerBound();
+  mlir::Value stopValue = op.getUpperBound();
+  mlir::Value stepValue = op.getStep();
+
+  auto startOp =
+      llvm::dyn_cast_or_null<arith::ConstantOp>(startValue.getDefiningOp());
+  auto stopOp =
+      llvm::dyn_cast_or_null<arith::ConstantOp>(stopValue.getDefiningOp());
+  auto stepOp =
+      llvm::dyn_cast_or_null<arith::ConstantOp>(stepValue.getDefiningOp());
+
+  if (!startOp || !stopOp || !stepOp) {
+    return StringError("only static loops can be analyzed");
+  }
+
+  auto startAttr = startOp.getValue().cast<mlir::IntegerAttr>();
+  auto stopAttr = stopOp.getValue().cast<mlir::IntegerAttr>();
+  auto stepAttr = stepOp.getValue().cast<mlir::IntegerAttr>();
+
+  if (!startOp || !stopOp || !stepOp) {
+    return StringError("only integer loops can be analyzed");
+  }
+
+  int64_t start = startAttr.getInt();
+  int64_t stop = stopAttr.getInt();
+  int64_t step = stepAttr.getInt();
+
+  result = calculateNumberOfIterations(start, stop, step);
+  return std::nullopt;
+}
+
+static std::optional<StringError> on_enter(scf::ForOp &op,
+                                           ExtractTFHEStatisticsPass &pass) {
+  int64_t numberOfIterations;
+
+  std::optional<StringError> error =
+      calculateNumberOfIterations(op, numberOfIterations);
+  if (error.has_value()) {
+    return error;
+  }
+
+  assert(numberOfIterations > 0);
+  pass.iterations *= (uint64_t)numberOfIterations;
+  return std::nullopt;
+}
+
+static std::optional<StringError> on_exit(scf::ForOp &op,
+                                          ExtractTFHEStatisticsPass &pass) {
+  int64_t numberOfIterations;
+
+  std::optional<StringError> error =
+      calculateNumberOfIterations(op, numberOfIterations);
+  if (error.has_value()) {
+    return error;
+  }
+
+  assert(numberOfIterations > 0);
+  pass.iterations /= (uint64_t)numberOfIterations;
+  return std::nullopt;
+}
+
+// ########
+// Dispatch
+// ########
+
+#define DISPATCH_ENTER(type)                                                   \
+  if (auto typedOp = llvm::dyn_cast<type>(op)) {                               \
+    std::optional<StringError> error = on_enter(typedOp, *this);               \
+    if (error.has_value()) {                                                   \
+      return error;                                                            \
+    }                                                                          \
+  }
+
+#define DISPATCH_EXIT(type)                                                    \
+  if (auto typedOp = llvm::dyn_cast<type>(op)) {                               \
+    std::optional<StringError> error = on_exit(typedOp, *this);                \
+    if (error.has_value()) {                                                   \
+      return error;                                                            \
+    }                                                                          \
+  }
+
+std::optional<StringError>
```

Comment:
we are used to use `llvm::Expected<>` here so that you can return the value or the error.

Reply:
Not sure how to replace the ones that doesn't return a value, can I do `llvm::Expected<void>`?